### DISCUSSION TITLE
Adding %json% variable for rules

### DISCUSSION THREAD

**Discussion by anishsane** [arendst/Tasmota#20709] [Ideas]
Consider a case of IR-receive.

The output is of the form:

    {"IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x02FD40BF","DataLSB":"0x40BF02FD","Repeat":0}}

If we want to send this to a logging server, we can use an mqtt broker.
Or we have to use rules snippets to concatenate the string, something like:

    on IrReceived#Protocol do var1 %value% endon on IrReceived#data do var2 %value% endon on IrReceived#datalsb do var3 %value% endon on IrReceived#bits do var4 %value% endon on IrReceived do webquery http://logger.local/irreecived/%var1%/%var2%/%var3%/%var4% GET endon 

Instead, it would be easier if we could use

    on IrReceived do webquery http://logger.local/irreecived POST %json% endon 

Of course, MQTT server is an option, but it is an overhead if your use case is limited.

(I tried searching if this suggestion is already covered, but did not find any.)

**Comment by barbudor** [comment]
Http is an overhead

If you are using an ESP32, you can do that in Berry

**Reply by anishsane** [reply]
No, I am using ESP8266-based ESP01 module.
The purpose of that particular device is to interpret the IR and send it to a server if it is a daikin216 protocol IR message.
So, I intended to use something like

    on IrReceived#Protocol=daikin216 do webquery http://another.iot.device/irreecived POST %json% endon 

I had this idea because it will also save on the rules size limit.

**Reply by sfromis** [reply]
In most cases, you just do not need all fields transmitted, with `Data` being the useful payload, and maybe filter for the `Protocol` value first.

**Reply by anishsane** [reply]
This is true.
But esp for AC IRrecv protocols, there are too many fields. So, we end up having to write many rules.

    RSL: RESULT = {"Time":"2019-09-09T21:52:35","IrReceived":{"Protocol":"PANASONIC_AC","Bits":216,"Data":"0x0220E004000000060220E00400032C805F06000EE0000081000089","Repeat":0,"IRHVAC":{"Vendor":"PANASONIC_AC","Model":2,"Power":"on","Mode":"auto","Celsius":"on","Temp":22,"FanSpeed":"medium","SwingV":"auto","SwingH":"middle","Quiet":"off","Turbo":"off","Econo":"off","Light":"off","Filter":"off","Clean":"off","Beep":"off","Sleep":-1}}}

The above line was copied from [tasmota docs](https://tasmota.github.io/docs/Tasmota-IR/#receiving-ir-commands). But the daikin216 protocol also has similar fields.

I could potentially use the data field and re-parse the raw data on the logger side. But it's sad to let go of the pre-parsed message.

For now, I will stick with mqtt based solution. I wanted to eliminate mqtt server because this is the only tasmota in the house that needs such logging.

**Reply by sfromis** [reply]
Using MQTT is not for logging, but as the preferable strategy for automation around Tasmota. If logging is your purpose, you may want the syslog feature.

**Reply by anishsane** [reply]
Sorry, I think I have used the wrong words and unnecessarily abstracted the things.

Let me explain the actual setup.

* I have a Raspberry Pi zero running a Python server to serve an AC remote control webpage over HTTP. We can use it to control the AC from the phone's browser.
* The actual IRsend operation is done by a Tasmota device with esp8266 that is controlled by the RPI over `http://tasmota-ip/cm?cmnd=...`
* I am also using IRRecv to update the Python server state in case we use the physical remote control to control the AC. That way, the browser based software remote is in sync with the updates from the physical remote.
* In that case, I want to send the IR receive data from Tasmota to the Python server via HTTP. Hence this proposal:

        on IrReceived#Protocol=daikin216 do webquery http://rpi.python.ip/irreecived POST %json% endon 

I am currently running a MQTT server that Tasmota pushes the IRrecv data to and then a monitor application that watches (subscribes) for these data updates on the MQTT server and updates the python server state.

**Reply by barbudor** [reply]
That was a cheap one : https://github.com/arendst/Tasmota/pull/20718
Wait for acceptance by Theo and merge

**Reply by anishsane** [reply]
Thank you so much!

**Reply by barbudor** [reply]
@anishsane Don't throw your MQTT broker to quickly
This change is not safe and make Tasmota crash as soon as the JSON is more than 10-15 characters
So definitively not a good solution

I have asked Theo to revert that change.
We will need more time to think is something more memory efficient is possible

**Reply by anishsane** [reply]
Hmm.
If it is a hardware limitation (due to memory size) we have to revert it.
Thanks for the efforts though!
